### ISSUE TITLE
Single-person profile page

### ISSUE THREAD

**Issue by richardolsson** [zetkin/app.zetkin.org#452]
Each person in the Zetkin database should have a "profile" page where organizers and other officials can interact with their data.

## Definitions of key terminology
* Person: a member/contact record in a Zetkin organization, managed by organization officials. Not quite the same as user.
* User: a user account in Zetkin, managed by Zetkin Foundation officials.
* Field: a piece of information on a Person record, e.g. their first name, email or phone number. There are some fields built into Zetkin, and other fields can be added to Person records by an organization using "custom fields"
* Tag: a person can be assigned any number of string-based tags, from a set of tags defined by the organization.

## Target audience
These are the roles that will most likely use this feature:

* Database managers/admins looking whose main job is to make sure that data is up to date and that irrelevant people are deleted from the database
* Organizers looking for information about a person

## Requirements (job stories)
Below are job stories that describe the requirements for this feature.

### Need to have
- [x] When I'm looking for information about a person, I want to be able to find everything on a single page / set of pages, so that I don't have to navigate to different parts of Zetkin
- [x] When I'm on a person's page, I want to see all their fields, so that I can easily find key pieces of information
- [x] When a piece of information in a field is incorrect, I want to be able to easily edit it so that I can keep the database clean
- [x] When a person wants to quit the organization, I want to delete them from the entire organization, so that no other sub-organizations can keep the record
- [x] When a person has moved away from my branch, I want to be able to remove them from my sub-organization only, so that other sub-organizations are unaffected
- [x] When I'm on a person's page, I want to see all tags that have been assigned to them, so that I can quickly get some context about them
- [x] When I want to group people or indicate some boolean state (e.g. paid yes/no) I want to be able to assign a tag to a person, so that it can be used elsewhere in Zetkin to filter, prioritize etc
- [x] When I'm assigning a tag to a person, I want to pick from a list of existing tags, so that I don't accidentally create the same tag twice
- [x] When there are a lot of tags to pick from, I want to be able search for a tag, so that I don't have to look through a huge list one by one
- [x] When a suitable tag does not yet exist, I want to be able to create the tag while I'm assigning it, so that I don't have to go somewhere else to find the tag

### Nice to have / known future improvements
- [ ] When I'm inspecting the data in a person's field, I want to be able to see when each of them last changed, so that I can get a hint of what might have become irrelevant
- [ ] When I create a tag, I want to be able to pick a color for that tag so that it can easily be discerned from (or grouped with) other tags visually
- [ ] When a person is connected to sub-organizations, I wan to be able to toggle whether to see information belonging to those organizations as well, so I can optionally get a complete picture of all the data stored on this person
- [ ] When I'm familiarizing myself with a person, I want to be able to see a timeline of all that they have done in the organization, so that I can know their history and level of experience
- [ ] When I have relevant information about a person that isn't suitable for a field or tag, I want to be able to write a free text note and have it timestamped, so that me and other organizers can remember it later
- [ ] When there are other records that appear to be duplicates of the one I'm looking at, I want to be alerted so that I can easily merge them and keep the database clean

## Degisns
https://www.figma.com/file/YFXcYN3LUhHghLb0cFqU85/Casework?node-id=1208%3A42080

## Solution
TBD

## Tasks
- [x] #554 
- [x] #555
- [x] #556 
- [x] #557
- [x] #558

**Comment by djbusstop** [comment]
Question: can/should we use in line text edit component

**Comment by djbusstop** [comment]
Once this work is done, we should see how we did the "cards" for displaying the grouped info, and try to come up with a generalisable design for displaying fields, and use it on tasks so we can have a consistent style.

**Comment by djbusstop** [comment]
@richardolsson @GemCopeland in the figma and wireframes there is an ellipsis menu, which we have on every other page, used mostly for opening a dialog to edit or delete the thing you're looking at. However there is no design with content for that menu.

For "edit" we are no longer using the manage tab, and the plan is for the manage tab to just handle deleting. But why break the existing pattern of putting a delete function in the ellipsis menu? Should we rather just do it that way?

**Comment by richardolsson** [comment]
> @richardolsson @GemCopeland in the figma and wireframes there is an ellipsis menu, which we have on every other page, used mostly for opening a dialog to edit or delete the thing you're looking at. However there is no design with content for that menu.
> 
> For "edit" we are no longer using the manage tab, and the plan is for the manage tab to just handle deleting. But why break the existing pattern of putting a delete function in the ellipsis menu? Should we rather just do it that way?

This separate section for deleting comes from my Christmas wireframes, and was inspired by the GitHub "Danger Zone". Compared to deleting anything else, deleting a person is a _very destructive_ action (because it also deletes, or at least makes unavailable, associated data).

On top of this, there are also two ways of deleting a person (from entire organization tree, or just from current sub-organization) which means it needs to be communicated clearly to the user. Hence the big special section for deleting.

That said, we could still use the ellipsis menu with two options for deleting, or a single option that triggers a modal for how to delete, both of which would then require confirmation. I'm open for either and would love to hear what both of you think.

**Comment by lowlandjuju** [comment]
I would advise not to offer deletion of a user as a standard UI option. Instead, replace the button with a **deactivate** button. You can then reserve a special delete function for specific requests from individuals that their data be deleted completely. And in that case you can still leave them in the database, but anonymise them...
You can even set up task runners to periodically, automatically anonymise user data after a given length of time, according to the privacy policy of the organisation.

If you do it like this then the 'delete' button becomes trivial and can be left in the UI as a simple button in the ellipsis menu.
Of course, there would need to be a little bit of extra UI to show/hide inactive members so you can resurrect them...

**Comment by GemCopeland** [comment]
> Once this work is done, we should see how we did the "cards" for displaying the grouped info, and try to come up with a generalisable design for displaying and editing fields, and use it on tasks so we can have a consistent style. Should we stick with edit dialogs triggered in the ellipsis menu, or inline editing?

I'll add this to my list!

**Comment by richardolsson** [comment]
Ok, let's rewind a little bit and talk about the fundamental question of delete vs deactivate, since it was brought up. I think there is an issue of terminology here, but it's also a very good idea.

> Instead, replace the button with a **deactivate** button. You can then reserve a special delete function for specific requests from individuals that their data be deleted completely. And in that case you can still leave them in the database, but anonymise them... You can even set up task runners to periodically, automatically anonymise user data after a given length of time, according to the privacy policy of the organisation.

This is actually close to what "delete" already does. It makes the person completely inaccessible to anyone via the API, so in that regard it's like actually deleting something. But behind the scenes, the object is just flagged and only finally deleted after a while, so that we can restore it for the organization upon request.

There is currently no way in the Zetkin API to inspect the "bin" of people (and other records) which have been flagged for deletion. So implementing it in Gen3 extends far beyond frontend. A feature like that in the API, along with interface for the organizations to use it, is something that we could build, but I reckon is outside the scope of this issue.

I suggest we move this discussion elsewhere (e.g. Slack) since it's a bigger change in direction than whether or not to have an ellipsis menu. It's a good one though! I personally have wanted to make actions less destructive for a while, and we've begun taking steps. Maybe this can be a good next step!

Coming back to the matter of "ellipsis menu" vs "danger zone", I don't really mind either personally. As long as a user can't just click to delete by accident, i.e. there are confirmation dialogs, I think either is fine. I also don't think we should change the "delete" terminology until we actually have a bin from which regular organization officials can restore data.

Another piece of terminology that I think could be improved is the "delete entirely" vs "delete from organization", but I think we should handle that over in #558 (will elaborate there now).

> I would advise not to offer deletion of a **user** as a standard UI option.

Lastly, I just want to point out (as I will whenever I get the chance, because I know this isn't obvious) that what we're working with here are not _users_, but a record that we call _person_, and they're not the same thing in Zetkin. An organization can have people in their database that do not also use Zetkin themselves (so not users), and a single Zetkin user can be represented by _person_ records in multiple organizations that the user is "connected to". Since users are beyond the scope of organizations, they can't at all be managed by organization officials, but only by us at Zetkin Foundation (and by the users themselves). No need to dive deeper into further details of this matter here, but anyone who wants to know more can contact me. 😊

**Comment by richardolsson** [comment]
I think realistically, since we don't really have a pinned down design for how to edit the fields on the person, the last item on this EPIC will need to be postponed until the next release.

I propose we move this EPIC and #557 to R33.

**Comment by richardolsson** [comment]
Issue #557 was solved earlier this year so this epic is now complete.

**Comment by joe-irving** [comment]
Some of these "nice to haves" are a block for one group I am supporting to use Zetkin (specifically, notes and timeline). I noticed they have not been added as issues. I am happy to open up an issue on this, but i have a feeling there is some stuff on process I am unclear on.
